ratis-server/src/main/java/org/apache/ratis/server/impl/TransactionManager.java

```diff
@@ -28,17 +29,17 @@
  * Managing {@link TransactionContext}.
  */
 class TransactionManager {
-  private final ConcurrentMap<Long, Supplier<TransactionContext>> contexts = new ConcurrentHashMap<>();
+  private final ConcurrentMap<TermIndex, Supplier<TransactionContext>> contexts = new ConcurrentHashMap<>();
 
-  TransactionContext get(long index) {
-    return Optional.ofNullable(contexts.get(index)).map(Supplier::get).orElse(null);
+  TransactionContext get(TermIndex termIndex) {
+    return Optional.ofNullable(contexts.get(termIndex)).map(Supplier::get).orElse(null);
   }
 
-  TransactionContext computeIfAbsent(long index, Supplier<TransactionContext> constructor) {
-    return contexts.computeIfAbsent(index, i -> constructor).get();
+  TransactionContext computeIfAbsent(TermIndex termIndex, Supplier<TransactionContext> constructor) {
+    return contexts.computeIfAbsent(termIndex, i -> constructor).get();
   }
 
-  void remove(long index) {
-    contexts.remove(index);
+  void remove(TermIndex termIndex) {
```

Comment:
A minor point, this will lead to some garbage because contexts linked to discarded log indexes may not get cleaned up.

Reply:
We call  `remove(..)` in `whenComplete(..)`; see https://github.com/apache/ratis/blob/24482561f98e9f9b0555dffbddd92111af303724/ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java#L1815

So, the context will be removed no matter the `stateMachineFuture` is completed normally or exceptionally.  Note that a `stateMachineFuture` will always be completed (unless for critical error that the server will be shut down).

Reply:
Thanks @szetszwo. I saw that, but I was wondering if that happens on followers (after handling `appendEntries`).

Reply:
```java
@@ -150,27 +151,27 @@ class PendingRequests {
       return permit;
     }
 
-    synchronized PendingRequest put(Permit permit, long index, PendingRequest p) {
-      LOG.debug("{}: PendingRequests.put {} -> {}", name, index, p);
+    synchronized PendingRequest put(Permit permit, PendingRequest p) {
+      LOG.debug("{}: PendingRequests.put {}", name, p);
       final Permit removed = permits.remove(permit);
       if (removed == null) {
         return null;
       }
       Preconditions.assertTrue(removed == permit);
-      final PendingRequest previous = map.put(index, p);
+      final PendingRequest previous = map.put(p.getTermIndex(), p);
       Preconditions.assertTrue(previous == null);
       return p;
     }
```
@duongkame , 

Before this change, `PendingRequests` uses index without term as the key.   If a follower put a different request with the same index, the `Preconditions.assertTrue(previous == null)` above will fail it.  We seem not seeing such failure.  We should test it.

Reply:
Got it, agree. It just always had the doubt and it's not really related to this change.

Reply:
> ... We seem not seeing such failure. ...

Oh, only leader will add pending requests.  You are right that followers may leave garbage in `TransactionManager` if the log is truncated.